mongoimport/common.go

```diff
@@ -399,28 +570,197 @@ func validateFields(fields []string) error {
 		if strings.Contains(field, "..") {
 			return fmt.Errorf("field '%v' cannot contain consecutive '.' characters", field)
 		}
-		// NOTE: since fields is sorted, this check ensures that no field
-		// is incompatible with another one that occurs further down the list.
-		// meant to prevent cases where we have fields like "a" and "a.c"
-		for _, latterField := range fieldsCopy[index+1:] {
-			// NOTE: this means we will not support imports that have fields that
-			// include e.g. a, a.b
-			if strings.HasPrefix(latterField, field+".") {
-				return fmt.Errorf("fields '%v' and '%v' are incompatible", field, latterField)
+	}
+	// This checks all other cases by building attempting to build a legal tree of fields.
+	// If useArrayIndexFields is set, the tree contains arrays of subtrees.
+	fieldTree := make(map[string]interface{})
+
+	for _, field := range inputFields {
+		fieldParts := strings.Split(field, ".")
+		_, err := addFieldToTree(fieldParts, 0, fieldTree, useArrayIndexFields)
+		if err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+// addFieldToTree is a recursive function that builds up a tree of fields. It is used to check
+// that fields are compatible with each other. When useArrayIndexFields is set, it is mutually recursive
+// with addFieldToArray(). It closely mimics the behaviour of setNestedDocumentValue(). See validateFields()
+// for more information on the validity checks that are made when constructing a tree of fields.
+func addFieldToTree(fieldParts []string, partIndex int, tree map[string]interface{}, useArrayIndexFields bool) (map[string]interface{}, error) {
+	value, exists := tree[fieldParts[partIndex]]
+
+	if exists {
+		if len(fieldParts[partIndex:]) == 1 {
```

Comment:
Yep, you're correct. Great catch!

Reply:
Except I think the code should read like: "given that head is terminal: if value is terminal it's a dup, otherwise it's incompatible". Otherwise you miss the case that head is terminal and it is incompatible with a subtree/array that already exists in the tree.